README
                               Jmx4Perl
                               ========

INTRODUCTION

  Jmx4Perl provides an alternate way for accessing Java JEE Server
  management interfaces which are based on JMX (Java Management
  Extensions). It is an agent based approach, where a small Java
  Webapplication deployed on the application server provides an
  HTTP/JSON based access to JMX MBeans registered within the
  application server.

HOW IT WORKS

  For the agent mode a small Java Agent WAR (web archive) needs to be
  deployed on the Java application server. This agent is provided by
  the Jolokia project (www.jolokia.org). There is no need to add any
  startup parameters to the application server and to open any
  additional ports. All communication takes places via HTTP where JSON
  objects are exchanged. Additionally, the agent benefits from the
  security infrastructure in place which every application server
  provides for web application. More information about the agent can
  be found at http://www.jolokia.org

  The Perl module JMX::Jmx4Perl accesses the deployed agent servlet
  and transform the request's results from JSON into a simple Perl
  object.

TOOLS

  This distribution comes with several tools, which uses the
  JMX::Jmx4Perl for accessing the server:

  jmx4perl          - Command line tool for gathering JMX information
  check_jmx4perl    - Full featured Nagios Plugin
  j4psh             - Interactive, readline based JMX shell with context 
                      sensitive command completion
  jolokia           - Utility for downloading and managing Jolokia 
                      agents.

INSTALLATION

  The Perl part installs as any other module via Module::Build, which
  you need to have installed. Using

  perl Build.PL
  ./Build installdeps             # If there are dependencies missing and you 
                                  # have Module::Build >= 0.36 installed.
  ./Build
  ./Build test
  ./Build install

  will install the modules. It is highly recommended to install the
  recommended dependent modules, too to get the full jmx4perl
  power. The set of 'required' modules is kept small and guarantees
  only that 'jmx4perl' and the modules around JMX::Jmx4Perl are
  working properly. The other tools (check_jmx4perl, j4psh and
  jolokia) require the recommended modules for proper working. Look
  into Build.PL for which tool requires which module.

  In order to download the Jolokia WAR agent into the local directory
  as jolokia.war, use the following command

      jolokia

  This agent "jolokia.war" needs to be deployed on the JEE Server to
  monitor. Please consult http://www.jolokia.org/agent.html for more
  information how to install the agent. E.g. for Tomcat this war file
  needs to be copied into the webapps directory.

  To test it, you can use 'jmx4perl' with the URL of the deployed
  agent:

      jmx4perl http://<jeeserver>:<port>/jolokia
      
  Consult 'man jmx4perl' for more information about this command
  utility.  
 
RESOURCES

  * Jmx4perl's source is hosted on github.com. You can clone the
    repository with git://github.com/rhuss/jmx4perl.git as URL

  * Interesting articles around Jmx4Perl, JMX and Nagios can be found
    at http://labs.consol.de Checkout the various post categories for
    selecting a specific topic.

  * https://metacpan.org/pod/distribution/jmx4perl is the canonical entry point for jmx4perl related
    information. 

NOTE 

  For you convenience, the latest Module::Build is included in this
  distribution, so there is no need of a locally install Module::Build
  for installing this suite. More information about Module::Build can
  be found http://search.cpan.org/~dagolden/Module-Build/

LICENSE
  
  Copyright (C) 2009-2011 Roland Huss (roland@cpan.org)

  Jmx4perl is free software: you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 2 of the License, or
  (at your option) any later version.

  jmx4perl is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with jmx4perl.  If not, see <http://www.gnu.org/licenses/>.

  A commercial license is available as well. You can either apply the
  GPL or obtain a commercial license for closed source
  development. Please contact roland@cpan.org for further information.

PROFESSIONAL SERVICES

  Just in case you need professional support for this module (or
  Nagios, JMX or JEE in general), you might want to have a look at
  http://www.consol.com/nagios-monitoring . Contact
  roland.huss@consol.de for further information (or use the contact
  form at http://www.consol.com/contact/ )

ACKNOWLEDGMENTS
  
  Big thanks go to ...

  * Gerhard Lausser, who initially pushed me to think harder
    about a better way for monitoring JEE Servers with Nagios.

  * Danijel Tasov for patching, patching, patching and keeping 
    an eye on contemporary perl styling.
 
  * All bug reporters and blog commenters for helping me to 
    increase the overall quality (and for letting me know that 
    this is not software for the ivory tower)

BUGS 

  Please report any bugs and/or feature requests at 
  http://rt.cpan.org/Public/Bug/Report.html?Queue=jmx4perl

AUTHOR

  roland@cpan.org


CHANGES
1.13 (2020-05-12)
  - Fixed bug in output for Tomcat 8 check_jmx4perl Nagios check
  - Changed CLI interface of 'jmx4perl list' and the possible arguments
    for LIST request. Note, that this change is not backwards compatible.
    However the fix is trivial: Instead of providing a full path into the
    LIST response as single argument (e.g. 'java.lang/type=Memory/HeapMemoryUsage')
    you have to provide two arguments for the MBean and the path into MBean's
    meta data (e.g. 'java.lang:type=Memory' and'HeapMemoryUsage')
  - Fixed Dockerfile
  - Add Tomcat 8 specific checks

1.12 (2015-07-28)

  - Configuration can be also a directory in wich case <dir>/jmx4perl.cfg is tried
    (e.g. ~/.j4p/jmx4perl.cfg)
  - Added Docker build
  - Fix boolean values to be strings "true"/"false" when deserialized.
  - Changed from "Nagios::Plugin" to "Monitoring::Plugin"

1.11 (2014-11-22)

  - When within a MultiCheck a single check causes an exception, the
    other checks are now still present in the output of check_jmx4perl
    and the overall check has the status UNKNWON (#40)
  - Minor fixes on the WebSphere Checks
  - Fixed issue when calling check_jmx4perl for an operation without argument (RT##98166)

1.10 (2014-06-30)

  - Added WebSphere checks

1.08 (2014-06-30)
  - Fixed warning when using MBeanName (#31)
  - Fixed BaseMBean and formatting
  - Fixed relative checks when using MBean patterns
  - Added formatter '%q' to include a ratio of value to base without
    multiplying by 100 like for '%r'
  - Disabled OpenPGPVerifier since it doesn't support the new digest
    algorithms used for signing the Jolokia artefacts (#32)
  - Don't set ssl_opts if on LWP < 6 (#28)
  - Fixed BaseMBean and BaseAttribute config directives (#25)
  - Fixed regexp for squeezing trailing slashes (RT#89108)
  - Fixed check definition for 'wls_channel_connections (RT#89107)
  - Changed check 'memory_gc_time' to be a relative check to measure the
    relative amount taken for the GC. If you use this check (or a sub-check of this)
    YOU NEED TO UPDATE YOUR THRESHOLDS (and regenerate the pn4p graphics) if you use this
    check directly
  - Fixed bug when using 0 thresholds in checks using parent checks (#38)
  - Added support for '*' wildcard when navigation with cd for j4psh
  - Fixed bug with check inheritance and check parameters which contain parantheses
  - Added an option "MultiCheckPrefix" for "Checks" in order to specify the prefix
    for multi checks
  - Added config options "SummaryOk" and "SummaryFailure" for allowing to fine tune
    multi check output (#24)
  - "Argument" can be used in "Operation" config checks for providing arguments to
    Nagios checks which are based on operations (#27)

1.07 (2013-04-16)
  - Added more robust timeout for the Jmx4Perl Agent (requires Sys::SigAction)
  - SSL Host key verification switched off when connecting via SSL
  - Fixed issue with quoting in j4psh (#14)
  - 'cat' in j4psh is now caseinsensitive when using wildcards (#18)
  - Added BaseMBean, BaseAttribute and BasePath as alternative to Base f
    or check_jmx4perl (#16)
  - "jolokia" can do 'repack' and 'info' also when not being connected
    to the internet (#20)
  - Multi-Checks can now reference other Multi-Checks either via <MultiCheck>
    or <Check> (#19)
  - Added new option '--perfdata true|false' (PerData false in configuration) for
    switching of performance data. Also, for string checks performance data is
    switched off always. (#22)
  - Added %y and %z as placeholder for configured CRITICAL and WARNING thresholds
    for the output provided with "Label" in check_jmx4perl (#13)

1.06 (2012-10-13)
  - A a scripting mode to check_jmx4perl which allows putting in arbitraty
    Perl code for extracting the value to match against
  - weblogic specific Nagios checks added
  - Added name as optional parameter for Nagios checks in tomcat.cfg (thanks Wolfgang)
  - When a multi checks fails, then the name of the check is added instead of its definition
    key. This allows for better direct usage of predefined checks in own multi checks.
  - If neither a CRITICAL nor a WARNING threshold is provided, then
    the check always returns OK. This is especially useful when the
    motivation is to only collect performance data.

1.05 (2012-04-22)
  - Added Time::HiRes as dependency to check_jmx4perl
  - Replaced XML::Tidy with XML::Twig and relaxed version number
    requirement on Module::Build
  - RT#72413: Fixed configuration in threads.cfg
  - Updated documentation for 'jmx4perl' (--method and --legacy-escape
    explained)
  - Bundled Module::Build 0.34 in order to improve the installation
    experience
  - j4psh: Added 'pwd' command
  - j4psh: Added options -a (attributes) and -o (operatiosn) to the 'ls'
    command which now also supports wildcards for filtering

1.04 (2011-11-27)
  - Fixed serious (and stupid) bug for jmx4perl and j4psh when printing
    out scalar values.

1.03 (2011-11-23)
  - Fixed stupid last minute bug.

1.02 (2011-11-23)
  - Fix for threshold with 0 value in check_jmx4perl
  - Fix automatic detection of the largest version number for Jolokia
    agents in with the format 1.0.1 when downloading with 'jolokia'
  - Fixed printing of boolean values for jmx4perl and j4psh for
    complex data structures (finally)
  - Added option '--option key=val' to jmx4perl and j4psh for tuning the
    output format of these tools (known keys: format,booleans,indent)
  - Added '--target' to j4psh so that it can operate against a JSR-160
    proxy

1.01 (2011-10-25)
  - Fixed 'jolokia' to load the new renamed jvm agent.
  - Fixed issue when printing boolean values with jmx4perl
  - Fixed issue with LWP as old as 5.805
  - Bumped required version of Module::Build to 0.38 in order to cope
    with messed up version number of XML::Tidy.
  - j4psh works now with  Getopt::Long before 2.38

1.00 (2011-10-3)
  - Changed escaping as introduced by Jolokia 1.0. If talking with

Build.PL
#!/usr/bin/perl 

# Check for Module::Build at the right version or use or own bundled one
# if the available one does not fit.
my $Minimal_MB = 0.34; 

my $Installed_MB =
  `$^X -e "eval q{require Module::Build; print Module::Build->VERSION} or exit 1"`;
chomp $Installed_MB;

$Installed_MB = 0 if $?;

# Use our bundled copy of Module::Build if it's newer than the installed.
unshift @INC, "inc/Module-Build" if $Minimal_MB > $Installed_MB;

require Module::Build;
use strict;
use Data::Dumper;

my %REQS = (
            "JSON" => "2.12",
            "LWP::UserAgent" => 0,
            "URI" => "1.35",
            "Data::Dumper" => 0,
            "Getopt::Long" => "2.38",
            "Carp" => 0,
            "Module::Find" => 0,
            "Scalar::Util" => 0,
            "base" => 0,
            "Sys::SigAction" => 0,
            "IO::Socket::Multicast" => 0 # opt
           );

my %SCRIPTS = ();

# Ask for various installation options:

print <<EOT;

Jmx4Perl comes with a set of supporting scripts, which 
are not necessarily required for using JMX::Jmx4Perl 
programmatically.
EOT

my $msg = <<EOT;

jmx4perl
========

jmx4perl is a command line utility for accessing Jolokia agents
(www.jolokia.org). It can be used for script based exploration 
and easy inspection of the JMX space.

Install 'jmx4perl' ? (y/n)
EOT

chomp $msg;
my $answer = y_n($msg,"y");
if ($answer) {
    add_reqs(
              "Crypt::Blowfish_PP" => 0 # opt
             );
    add_script("scripts/jmx4perl" => 1);   
}

my $msg = <<EOT;

check_jmx4perl
==============

check_jmx4perl is a full featured Nagios Plugin (www.nagios.org) for 
monitoring JEE and other Java-servers.

Install 'check_jmx4perl' ? (y/n)
EOT

chomp $msg;
my $answer = y_n($msg,"y");
if ($answer) {
    add_reqs(
              "Monitoring::Plugin" => "0.37", # req
              "Text::ParseWords" => 0,    # req
              "Time::HiRes" => 0,         # req
              "Config::General" => "2.34",# req
              "Pod::Usage" => 0,          # opt
              "Crypt::Blowfish_PP" => 0   # opt
             );
    add_script("scripts/check_jmx4perl" => 1);   
}

$msg = <<EOT;

cacti_jmx4perl
==============

cacti_jmx4perl is a script which can be used as a Cacti 
(www.cacti.net) plugin.

Install 'cacti_jmx4perl' ? (y/n)
EOT
chomp $msg;
$answer = y_n($msg,"y");
if ($answer) {
    add_reqs(
              "Monitoring::Plugin" => "0.37", # req
              "Text::ParseWords" => 0,    # req
              "Config::General" => "2.34",# req
              "Pod::Usage" => 0,          # opt
              "Crypt::Blowfish_PP" => 0 # opt
             );
    add_script("scripts/cacti_jmx4perl" => 1);   
}

$msg = <<EOT;

j4psh
=====

j4psh is an interactive JMX shell with context sensitive command line
completion. It uses JMX::Jmx4Perl for connecting to the JMX backend
and has quite some Perl module dependencies.

Install 'j4psh' ? (y/n)
EOT
chomp $msg;
$answer = y_n($msg,"y");

if ($answer) {
    add_reqs(
              "Getopt::Long" => 0,  # req, GetOptionsFromArray must be exported
              "Term::ShellUI" => 0,      # req
              "Term::Clui" => 0,         # req 
              "Term::Size" => "0.207",   # opt 
              "Config::General" => "2.34",# opt
              "File::SearchPath" => 0,   # opt 
              "Crypt::Blowfish_PP" => 0  # opt
             );
    add_script("scripts/j4psh" => 1);
    # check for Term::ReadLine::Gnu
    my $has_gnu_readline = eval "require Term::ReadLine; require Term::ReadLine::Gnu; 1";
    my $has_perl_readline = eval "require Term::ReadLine::Perl; 1";
    if (!$has_gnu_readline) {
        $msg = <<EOT;

Term::ReadLine::Gnu is the recommended readline module, but it is not
necessarily required. It needs a development variant of libreadline
installed along with header files.

Use Term::ReadLine::Gnu ? (y/n)
EOT
        chomp $msg;
        $answer = y_n($msg,"n");
        if ($answer) {
            add_reqs("Term::ReadLine::Gnu" => 0);
        } elsif (!$has_perl_readline) {
            add_reqs("Term::ReadLine::Perl" => 0,
                      "Term::ReadKey" => 0);
        }
    }
}

$msg = <<EOT;

jolokia
=======

jolokia is an utility which helps in downloading
and managing the Jolokia agents (www.jolokia.org), which 
are required on the server side for using jmx4perl.

Install 'jolokia' ? (y/n)
EOT
chomp $msg;
$answer = y_n($msg,"y");
if ($answer) {
    add_reqs(
              "Archive::Zip" => 0,        # req
              "XML::LibXML" => 0,         # req
              "File::Temp" => 0,          # req
              "Digest::MD5" => 0,         # opt
              "Digest::SHA1" => 0,        # opt
              "XML::Twig" => 0,           # opt
              "Term::ProgressBar" => 0    # opt
             );
    add_script("scripts/jolokia" => 1);
    my $has_openpgp = eval "require Crypt::OpenPGP; 1";
    if (!$has_openpgp) {
        my $check = `gpg --version`;
        if ($?) {
            $check = `gpg2 --version`;
            if ($?) {
                $msg = <<EOT;
                
jolokia uses PGP verification for the files downloaded, but neither
Crypt::OpenPGP nor GnuPG is installed. It is highly recommended to 
install at least one of them. Installing Crypt::OpenPGP however can 
be a pain due to its large set of dependencies.

Use Crypt::OpenPGP ? (y/n)
EOT
                chomp $msg;
                $answer = y_n($msg,"y");
                if ($answer) {
                    add_reqs("Crypt::OpenPGP" => 0);
                }
            }
        }
    }
}

# Add extra requirements
sub add_reqs {
    my %to_add = @_;
    for my $k (keys %to_add) {
        $REQS{$k} = $to_add{$k};
    }
}

sub add_script {
    my $script = shift;
    $SCRIPTS{$script} = 1;
}

sub y_n { 
    Module::Build->y_n(@_);
}

# ================================================================================ 

my $build = Module::Build->new
  (
   dist_name => "jmx4perl",
   dist_version_from => "lib/JMX/Jmx4Perl.pm",
   dist_author => 'Roland Huss (roland@cpan.org)',
   dist_abstract => 'Easy JMX access to Java EE applications',
   #sign => 1,
   installdirs => 'site',
   license => 'gpl',
   
   requires => \%REQS,
   script_files => \%SCRIPTS,

   build_requires => {
                      "Module::Build" => "0.34",
                      "Test::More" => "0",
                     },                       
   configure_requires => { 'Module::Build' => 0.34 },
   keywords => [  "JMX", "JEE", "Management", "Nagios", "Java", "Jolokia", "OSGi", "Mule" ],
  );

$build->create_build_script;

# ===================================================================================

